keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/ArtifactRepositoryTests.kt

```diff
@@ -0,0 +1,111 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.api.ArtifactType.DEB
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryArtifactVersion
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import strikt.api.Assertion
+import strikt.api.expectThat
+import strikt.api.expectThrows
+import strikt.assertions.first
+import strikt.assertions.hasSize
+import strikt.assertions.isEqualTo
+import java.net.URI
+
+abstract class ArtifactRepositoryTests<T : ArtifactRepository> : JUnit5Minutests {
+  abstract fun factory(): T
+
+  open fun flush() {}
+
+  data class Fixture<T : ArtifactRepository>(
+    val artifact: DeliveryArtifact,
+    val repository: T
+  )
+
+  fun tests() = rootContext<Fixture<T>> {
+    fixture {
+      Fixture(
+        artifact = DeliveryArtifact("fnord", DEB),
+        repository = factory()
+      )
+    }
+
+    before {
+      repository.store(artifact)
+    }
+
+    after {
+      flush()
+    }
+
+    context("registering a new artifact version") {
+      context("the artifact is unknown") {
+        test("registering a new version throws an exception") {
+          expectThrows<IllegalArgumentException> {
+            repository.store(
+              DeliveryArtifactVersion(
+                DeliveryArtifact("some-other-artifact", DEB),
+                "58.0",
+                URI("https://my.jenkins.master/job/some-other-artifact/58")
+              )
+            )
+          }
+        }
+      }
+
+      context("the artifact version already exists") {
+        before {
+          repeat(2) {
+            repository.store(
+              DeliveryArtifactVersion(
+                artifact,
+                "1.0",
+                URI("https://my.jenkins.master/job/${artifact.name}-release/1")
+              )
+            )
+          }
+        }
+
+        // TODO: should we throw an exception instead?
```

Comment:
I don't think so, because we may get the same message more than once, and that should be nicely handled

Reply:
Yeah, agreed